pydicom/data/download.py

```diff
@@ -51,53 +65,110 @@ def calculate_file_hash(filename):
     return hasher.hexdigest()
 
 
-def get_config_dir():
+def get_config_dir() -> str:
+    """Return the path to the pydicom config directory, creating it if required
+
+    The config directory will be named ``.pydicom`` and will be created in the
+    local user's home directory.
+    """
     config_dir = pathlib.Path.home().joinpath(".pydicom")
     config_dir.mkdir(exist_ok=True)
 
     return config_dir
 
 
 @retry.retry(urllib.error.HTTPError)
-def download_with_progress(url, filepath):
+def download_with_progress(url: str, fpath: pathlib.Path) -> None:
+    """Download the file at `url` to `fpath` with a progress bar.
+
+    Parameters
+    ----------
+    url : str
+        The URL to download the file from.
+    fpath : pathlib.Path
+        The absolute path where the file will be written to.
+    """
     with DownloadProgressBar(
         unit="B", unit_scale=True, miniters=1, desc=url.split("/")[-1]
     ) as t:
-        urllib.request.urlretrieve(url, str(filepath), reporthook=t.update_to)
+        urllib.request.urlretrieve(
+            url, os.fspath(fpath), reporthook=t.update_to
+        )
 
 
-def get_data_dir():
+def get_data_dir() -> str:
+    """Return the path to the cache directory, creating it if required."""
     data_dir = get_config_dir().joinpath("data")
     data_dir.mkdir(exist_ok=True)
 
     return data_dir
 
 
 @functools.lru_cache()
-def get_url_map():
-    with open(str(HERE.joinpath("urls.json")), "r") as url_file:
-        url_map = json.load(url_file)
-
-    return url_map
-
-
-def get_url(filename):
-    url_map = get_url_map()
-
+def get_url_map() -> Dict[str, str]:
+    """Return a dict containing the URL mappings from ``urls.json```."""
+    with open(HERE.joinpath("urls.json"), "r") as url_file:
+        return json.load(url_file)
+
+
+def get_url(filename: str) -> str:
+    """Return the download URL corresponding to `filename`.
+
+    The filename:URL mappings are located in the ``urls.json`` file.
+
+    Parameters
+    ----------
+    filename : str
+        The filename of the file to get the corresponding URL for.
+
+    Returns
+    -------
+    str
+        The download URL corresponding to `filename`.
+
+    Raises
+    ------
+    ValueError
+        If `filename` is not in the ``urls.json`` record.
+    """
+    # Convert filename to lowercase because windows filenames are
+    #   case-insensitive
+    urls = {k.lower(): v for k, v in get_url_map().items()}
```

Comment:
@SimonBiggs you might be interested in the case modifications for urls.json and hashes.json. I was running into issues on Windows with `Path.glob()` randomly (?) returning lowercase versions of the filenames.

In any case, we need to check that the filenames in urls.json and hashes.json are unique accounting for case-insensitivity

Reply:
Yup, I like that solution. That makes sense. Thanks for the heads up :).